src/typecheck/eq.rs

```diff
@@ -0,0 +1,356 @@
+        // annotations. Documentation and merge priority shouldn't impact the result on the
+        // other hand.
+        (MetaValue(m1), MetaValue(m2)) => {
+            let value_eq = match (&m1.value, &m2.value) {
+                (None, None) => true,
+                (Some(v1), Some(v2)) => contract_eq_bounded(state, v1, env1, v2, env2),
+                _ => false,
+            };
+
+            // We use the same logic as in the typechecker: the type associated to an annotated
+            // value is either the type annotation, or the first contract annotation.
+            let ty1 = m1.types.as_ref().or_else(|| m1.contracts.first());
+            let ty2 = m2.types.as_ref().or_else(|| m2.contracts.first());
+
+            let ty_eq = match (ty1, ty2) {
+                (None, None) => true,
+                (Some(ctr1), Some(ctr2)) => type_eq_bounded(
+                    state,
+                    &TypeWrapper::from(ctr1.types.clone()),
+                    env1,
+                    &TypeWrapper::from(ctr2.types.clone()),
+                    env2,
+                ),
+                _ => false,
+            };
+
+            value_eq && ty_eq
+        }
+        // We dont treat imports and parse errors
+        _ => false,
+    }
+}
+
+/// Compute the equality between two hashmaps holding either types or terms.
+fn map_eq<V, F>(
+    mut f: F,
+    state: &mut State,
+    map1: &HashMap<Ident, V>,
+    env1: &TermEnvironment,
+    map2: &HashMap<Ident, V>,
+    env2: &TermEnvironment,
+) -> bool
+where
+    F: FnMut(&mut State, &V, &TermEnvironment, &V, &TermEnvironment) -> bool,
+{
+    map1.len() == map2.len()
+        && map1.iter().all(|(id, v1)| {
+            map2.get(id)
+                .map(|v2| f(state, v1, env1, v2, env2))
+                .unwrap_or(false)
+        })
+}
+
+/// Convert rows (represented as linked list for the needs of unification) to a hashmap to perform
+/// the type equality comparison.
+///
+/// Require the rows to be closed (i.e. the last element must be RowEmpty), otherwise `None` is
+/// returned. `None` is returned as well if a type encountered is not a row type, a if it is a enum
+/// row.
+fn rows_as_map(ty1: &TypeWrapper) -> Option<HashMap<Ident, &TypeWrapper>> {
+    let mut map = HashMap::new();
+    let mut curr = ty1;
+
+    loop {
+        match curr {
+            TypeWrapper::Concrete(AbsType::RowEmpty()) => return Some(map),
+            TypeWrapper::Concrete(AbsType::RowExtend(id, Some(ty), tail)) => {
+                map.insert(id.clone(), ty);
+                curr = tail.as_ref();
+            }
+            _ => return None,
+        }
+    }
+}
+
+/// Perform the type equality comparison on types. Structurally recurse into type constructors and test
+/// that subtypes or subterms (contracts) are equals.
+///
+/// Currently, this function morally operates on `Types` rather than `TypeWrapper`s as it is called
```

Comment:
`morally`?

Reply:
In the sense that the official type is `TypeWrapper`, but "in fact", "actually", it is operating on the subset `Types`. Is this use of `morally` strange, uncommon or totally invalid? I've used it for a long time, but it wouldn't be the first time one discovers they've been done something wrong their whole life :sweat_smile:

Reply:
I have not seen it used this way. Feel free to resolve this conversation anyway.